### PR TITLE
PubSub: Prevent unhandled background error on SPM shutdown

### LINKED ISSUE #6130: PubSub: dispatch_callback errors after cancel() of a subscription
OS: Linux 4.17.0-3rodete2-amd64, Debian 4.17.17-1rodete2
Python 2.7.13
google-cloud-pubsub 0.35.4

## Stacktrace 1
```
ERROR:google.cloud.pubsub_v1.subscriber._protocol.helper_threads:Error in queue callback worker: 'NoneType' object has no attribute 'is_paused'
Traceback (most recent call last):
  File "/usr/local/google/home/ehudm/virtualenvs/beamenv/local/lib/python2.7/site-packages/google/cloud/pubsub_v1/subscriber/_protocol/helper_threads.py", line 112, in __call__
    self._callback(items)
  File "/usr/local/google/home/ehudm/virtualenvs/beamenv/local/lib/python2.7/site-packages/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py", line 101, in dispatch_callback
    self.ack(batched_commands.pop(requests.AckRequest))
  File "/usr/local/google/home/ehudm/virtualenvs/beamenv/local/lib/python2.7/site-packages/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py", line 124, in ack
    self.drop(items)
  File "/usr/local/google/home/ehudm/virtualenvs/beamenv/local/lib/python2.7/site-packages/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py", line 133, in drop
    self._manager.maybe_resume_consumer()
  File "/usr/local/google/home/ehudm/virtualenvs/beamenv/local/lib/python2.7/site-packages/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py", line 224, in maybe_resume_consumer
    if not self._consumer.is_paused:
AttributeError: 'NoneType' object has no attribute 'is_paused'
```

## Stacktrace 2
```
ERROR:google.cloud.pubsub_v1.subscriber._protocol.helper_threads:Error in queue callback worker: 'NoneType' object has no attribute 'remove'
Traceback (most recent call last):
  File "/usr/local/google/home/ehudm/virtualenvs/beamenv/local/lib/python2.7/site-packages/google/cloud/pubsub_v1/subscriber/_protocol/helper_threads.py", line 112, in __call__
    self._callback(items)
  File "/usr/local/google/home/ehudm/virtualenvs/beamenv/local/lib/python2.7/site-packages/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py", line 101, in dispatch_callback
    self.ack(batched_commands.pop(requests.AckRequest))
  File "/usr/local/google/home/ehudm/virtualenvs/beamenv/local/lib/python2.7/site-packages/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py", line 124, in ack
    self.drop(items)
  File "/usr/local/google/home/ehudm/virtualenvs/beamenv/local/lib/python2.7/site-packages/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py", line 132, in drop
    self._manager.leaser.remove(items)
AttributeError: 'NoneType' object has no attribute 'remove'
```

## Steps to reproduce
```py
event = threading.Event()
def message_cb(msg):
  msg.ack()
  event.set()
sub_client = pubsub.SubscriberClient()
sub_future = sub_client.subscribe(sub_name, message_cb)
event.wait()
sub_future.cancel()
```

This subscription should be canceled approximately after the first message. In my test there are two messages sent to the topic and `event.set()` is called after the second.

I realize that calling cancel() so soon prevents the ack from being sent. Is there a cleaner way to stop the subscription?

### PR DESCRIPTION
**Pull request by plamut**
Closes #6130.

This PR fixes errors in background threads that might happen on user-initiated streaming pull shutdown, i.e. cancelling the future returned by the `subscriber.subscribe()` call.

### How to test

**Steps to reproduce:**
- Follow the steps in the [issue description](https://github.com/googleapis/google-cloud-python/issues/6130#issue-364719645).
- Preferably set the log level of the test script to `logging.DEBUG`. FWIW, I used the following config during development:
```py
import logging

log_format = (
    "%(levelname)-8s [%(asctime)s] %(threadName)-33s "
    "[%(name)s] [%(filename)s:%(lineno)d][%(funcName)s] %(message)s"
)
logging.basicConfig(level=logging.DEBUG, format=log_format)
```

**Actual result (before the fix):**
An error can be observed in `Thread-CallbackRequestDispatcher` (the one listed under "Stacktrace 2" in the issue description). The issue is reproducible with high consistency:
```
AttributeError: 'NoneType' object has no attribute 'remove'
```
> NOTE: The error listed under "Stacktrace 1" is **not** reproducible anymore, it has been fixed quite some time ago.

**Expected result (after the fix):**
The StreamingPullManager gets shut down cleanly, i.e. without the error reported. There might be a stacktrace of the `Cancelled` exception produced by `api_core.bidi`, but that is a harmless "expected exception":
```
Thread-ConsumeBidirectionalStream caught error 499 Locally cancelled by application! and will exit. Generally this is due to the RPC itself being cancelled and the error will be surfaced to the calling code
...
google.api_core.exceptions.Cancelled: 499 Locally cancelled by application!
```

### CODE REVIEW

File: pubsub/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py [hunk 1]
```diff
@@ -125,17 +125,31 @@ def drop(self, items):
...
-        self._manager.maybe_resume_consumer()
+        # If the manager is in the process of being shut down, the leaser might
+        # not exist on it anymore, thus we need to obtain our own reference to
+        # it and check it for None to avoid errors.
+        # Nevertheless, calling maybe_resume_consumer() in the manager shutdown
+        # state is still fine, as it is effectively a no-op at that point.
+        leaser = getattr(self._manager, "leaser", None)
+        if leaser is not None:
+            leaser.remove(items)
+            self._manager.maybe_resume_consumer()
```
Comment (tseaver) [inline pubsub/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py]: This seems racy:  should it be protected by the `_closing` lock on the manager?

Comment (plamut) [reply pubsub/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py]: You mean the `maybe_resume_consumer()` method or the entire `dispatcher.drop()` method?

It is _likely_ to be fine based on the logs and interaction between the threads, however, "likely" is both not a strong enough word, and it's too dependent on the actual implementation. Having an explicit lock wold make it easier to reason about the code without relying on implicit assumptions, will add it.

(as for the leaser add() / remove() methods, that [other PR](https://github.com/googleapis/google-cloud-python/pull/7948/commits/9b010605f0796a73f3a0e65bee4cba9c6727238e) handles that, will respond there tomorrow).

**Comment by plamut** [comment]
The main problem here is that the leaser and dispatcher use each other's methods, thus cannot be shut down in a clear linear sequence (and breaking this inter-dependency would likely require substantial refactoring).

It's also tricky to synchronize the manager shutdown with any remaining work that dispatcher might have left in its queue (such as pending ACK requests). The thread that initiated the manager shutdown waits on the dispatcher thread to terminate, thus if the latter tries to sync with the manager shutdown through a shared lock, a deadlock can happen.

The current fix uses the fact that once in the shutdown mode, several operations on the manager have no effect (`maybe_resume_consumer()` / `maybe_pause_consumer()`), as do not the `add()` and `remove()` operations on a stopped leaser.

I still need to think of a meaningful yet understandable test to properly cover this.

**Comment by tseaver** [comment]
Could we delay setting the attributes to `None` until the very end?

**Comment by plamut** [comment]
@tseaver Definitely, that's exactly the fix that I am just about to submit. Much easier and cleaner than trying to refactor the streaming pull and trying to break the cyclic dependency between the dispatcher and leaser.

**Pull request merged**